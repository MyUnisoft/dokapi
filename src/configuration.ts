// Import Node.js Dependencies
import path from "path";
import fs from "fs/promises";

// Import Third-party Dependencies
import Ajv from "ajv";
import TOML from "@iarna/toml";

// Import Internal Dependencies
import schema from "./schema/configuration.schema.json";

// CONSTANTS
const kSchemaValidator = new Ajv();

export interface DokapiMenu {
  /** Title of the page (Will be extracted from markdown if not provided) */
  title?: string;

  /** System path to page or the directory containing the page */
  path: string;

  /** Is the menu collapsable (see stripe API for an example) */
  collapsable?: boolean;

  menu?: DokapiSubMenu[];
}

export type DokapiSubMenu = Omit<DokapiMenu, "menu" | "collapsable">;

export interface DokapiConfiguration {
  /** Title of the API */
  title: string;

  /** The logo to show under the title */
  logo?: string;

  /** The list of accessible menu (pages) on the left */
  menu: DokapiMenu[];
}

export type DokapiSubConfiguration = Pick<DokapiConfiguration, "title"> & {
  menu: DokapiSubMenu[]
};

/**
 * @description Read, parse and validate a Dokapi configuration.
 */
export async function read(location: string): Promise<DokapiConfiguration> {
  const str = await fs.readFile(
    path.join(location, "dokapi.config.toml"), "utf-8"
  );
  const manifest = TOML.parse(str);

  const validate = kSchemaValidator.compile(schema);
  if (!validate(manifest)) {
    console.log(validate.errors);

    throw new Error("Unable to validate Dokapi configuration");
  }

  return manifest as unknown as DokapiConfiguration;
}
