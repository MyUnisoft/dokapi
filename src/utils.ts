/**
 * @description Search for a title in a given Markdown string
 * @example
 * searchMarkdownHeadTitle("# hello world"); // hello world
 */
export function searchMarkdownHeadTitle(markdown: string): string | null {
  const result = /#(.*)/gm.exec(markdown);

  if (result !== null) {
    return result[1].trim();
  }

  return null;
}
