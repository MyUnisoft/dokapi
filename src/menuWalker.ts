// Import Internals
import * as DokapiConfig from "./configuration";

export async function* walkDokapiMenu(location: string, menuList: DokapiConfig.DokapiMenu[]) {
  for (const { title, path } of menuList) {
    console.log(title, path);

    yield "foobar";
  }
}
