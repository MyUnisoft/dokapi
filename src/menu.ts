// Import Node.js Dependencies
import path from "path";
import fs from "fs/promises";

// Import Third-party Dependencies
import MarkdownIt from "markdown-it";

// Import Internals
import * as DokapiConfig from "./configuration";
import * as utils from "./utils";

// VARS & CONSTANTS
const md = new MarkdownIt();

export type ParsedMenu = Pick<DokapiConfig.DokapiMenu, "title" | "collapsable"> & {
  /** HTML content (parsed from the menu markdown) */
  html: string;

  subMenu?: boolean;
};

export async function* walk(
  location: string,
  menuList: DokapiConfig.DokapiMenu[],
  subMenu = false
): AsyncIterableIterator<ParsedMenu> {
  for (const menu of menuList) {
    const menuPath = path.join(location, menu.path);
    const stat = await fs.stat(menuPath);

    const readmePath = stat.isDirectory() ? path.join(menuPath, "README.md") : menuPath;
    const readmeStr = await fs.readFile(readmePath, "utf-8");

    const title = menu.title ?? utils.searchMarkdownHeadTitle(readmeStr);
    if (typeof title === "undefined" || title === null) {
      throw new Error(`Unable to found a title for path: ${menuPath}`);
    }

    yield {
      title,
      collapsable: menu.collapsable ?? false,
      subMenu,
      html: md.render(readmeStr)
    };

    if (Array.isArray(menu.menu)) {
      yield* walk(stat.isDirectory() ? menuPath : path.dirname(menuPath), menu.menu, true);
    }
  }
}
