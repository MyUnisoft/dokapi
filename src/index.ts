// Import Node.js Dependencies
// import path from "path";
// import fs from "fs/promises";

// Import Internals
import * as DokapiConfig from "./configuration";
import { walkDokapiMenu } from "./menuWalker";

export interface generateDocumentationOptions {
  /** Default: `docs` */
  outDir?: string;
}

export async function generateDocumentation(location: string, options: generateDocumentationOptions = {}) {
  const { outDir = "docs" } = options;
  console.log(outDir);

  const config = await DokapiConfig.read(location);
  for await (const any of walkDokapiMenu(location, config.menu)) {
    console.log(any);
  }
}
