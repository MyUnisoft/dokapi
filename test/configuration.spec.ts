/* eslint-disable arrow-body-style */

// Import Node.js Dependencies
import path from "path";

// Import Internal Dependencies
import * as DokapiConfig from "../src/configuration";

// CONSTANTS
const kFixturesPath = path.join(__dirname, "fixtures");

describe("read", () => {
  it("should be able to READ a valid dokaconfig.config.toml", async() => {
    const result = await DokapiConfig.read(
      path.join(kFixturesPath, "valid")
    );

    expect(result).toMatchSnapshot("ValidDokapiConfig");
  });

  it("should throw an error for an invalid configuration", async() => {
    expect(async() => {
      return await DokapiConfig.read(path.join(kFixturesPath, "invalid"));
    }).rejects.toThrow("Unable to validate Dokapi configuration, Reason: must have required property 'menu'");
  });

  it("should throw an error if the provided location doesn't exist", async() => {
    expect(async() => {
      return await DokapiConfig.read(path.join(kFixturesPath, "edklblelle"));
    }).rejects.toThrow();
  });
});
