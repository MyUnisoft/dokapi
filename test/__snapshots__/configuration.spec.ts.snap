// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read should be able to READ a valid dokaconfig.config.toml: ValidDokapiConfig 1`] = `
Object {
  "menu": Array [
    Object {
      "path": "introduction",
      "title": "Hello world",
    },
    Object {
      "collapsable": true,
      "path": "products.MD",
    },
  ],
  "title": "MyUnisoft Partners API",
}
`;
